### PR TITLE
WosRecord - utility for working with a WoS <REC> XML element

### PR DESCRIPTION
**Pull request by darrenleeweber**
Connected to #265 
Connected to #274 

This could be a useful wrapper on a WoS `<REC>` element.  It will help in mapping record data into a PubHash.  It might also become a base class for other classes that provide more specific data accessors when a REC is from a WOS, MEDLINE, etc. database.

The `WosRecord#summary` (and supporting methods) convert the <REC> XML data into a Hash, which is further wrapped in an OpenStruct by `WosRecord#summary_struct`.  It must be noted that the Nokogiri doc does not automatically convert to JSON, e.g. `doc.to_json => "[]"`.

For example, this is the example fixture <REC> summary data and the conversions:
```xml
  <static_data>
    <summary>
      <EWUID>
        <WUID coll_id='WOS'/>
        <edition value='WOS.SSCI'/>
        <edition value='WOS.SCI'/>
      </EWUID>
      <pub_info issue='5' pubtype='Journal' sortdate='1972-01-01' has_abstract='N' coverdate='1972' vol='33' pubyear='1972'>
        <page end='413' page_count='1' begin='413'>413-413</page>
      </pub_info>
      <titles count='6'>
        <title type='source'>COLLEGE &amp; RESEARCH LIBRARIES</title>
        <title type='source_abbrev'>COLL RES LIBR</title>
        <title type='abbrev_iso'>Coll. Res. Libr.</title>
        <title type='abbrev_11'>COLL RES LI</title>
        <title type='abbrev_29'>COLL RES LIBR</title>
        <title type='item'>
          LIBRARY MANAGEMENT - BEHAVIOR-BASED PERSONNEL SYSTEMS (BBPS) -
          FRAMEWORK FOR ANALYSIS - KEMPER,RE
        </title>
      </titles>
      <names count='1'>
        <name daisng_id='19669717' seq_no='1' role='author'>
          <display_name>WEBER, DC</display_name>
          <full_name>WEBER, DC</full_name>
          <wos_standard>WEBER, DC</wos_standard>
          <first_name>DC</first_name>
          <last_name>WEBER</last_name>
        </name>
      </names>
      <doctypes count='1'>
        <doctype>Book Review</doctype>
      </doctypes>
      <publishers>
        <publisher>
          <address_spec addr_no='1'>
            <full_address>50 E HURON ST, CHICAGO, IL 60611</full_address>
            <city>CHICAGO</city>
          </address_spec>
          <names count='1'>
            <name addr_no='1' role='publisher' seq_no='1'>
              <display_name>ASSOC COLL RESEARCH LIBRARIES</display_name>
              <full_name>ASSOC COLL RESEARCH LIBRARIES</full_name>
            </name>
          </names>
        </publisher>
      </publishers>
    </summary>
    <!-- snipped out the <fullrecord_metadata> element -->
  </static_data>
```

```ruby
{"doctypes"=>["Book Review"],
 "names"=>
  [{"daisng_id"=>"19669717",
    "seq_no"=>"1",
    "role"=>"author",
    "display_name"=>"WEBER, DC",
    "full_name"=>"WEBER, DC",
    "wos_standard"=>"WEBER, DC",
    "first_name"=>"DC",
    "last_name"=>"WEBER"}],
 "pub_info"=>
  {"issue"=>"5",
   "pubtype"=>"Journal",
   "sortdate"=>"1972-01-01",
   "has_abstract"=>"N",
   "coverdate"=>"1972",
   "vol"=>"33",
   "pubyear"=>"1972",
   "page"=>{"end"=>"413", "page_count"=>"1", "begin"=>"413"}},
 "publishers"=>
  [{"addr_no"=>"1",
    "role"=>"publisher",
    "seq_no"=>"1",
    "display_name"=>"ASSOC COLL RESEARCH LIBRARIES",
    "full_name"=>"ASSOC COLL RESEARCH LIBRARIES",
    "address"=>{"addr_no"=>"1", "full_address"=>"50 E HURON ST, CHICAGO, IL 60611", "city"=>"CHICAGO"}}],
 "titles"=>
  {"source"=>"COLLEGE & RESEARCH LIBRARIES",
   "source_abbrev"=>"COLL RES LIBR",
   "abbrev_iso"=>"Coll. Res. Libr.",
   "abbrev_11"=>"COLL RES LI",
   "abbrev_29"=>"COLL RES LIBR",
   "item"=>"LIBRARY MANAGEMENT - BEHAVIOR-BASED PERSONNEL SYSTEMS (BBPS) - FRAMEWORK FOR ANALYSIS - KEMPER,RE"}}
```

The summary Hash is easily converted to JSON using `WosRecord#summary.to_json`, e.g.
```json
"{\"doctypes\":[\"Book Review\"],\"names\":[{\"daisng_id\":\"19669717\",\"seq_no\":\"1\",\"role\":\"author\",\"display_name\":\"WEBER, DC\",\"full_name\":\"WEBER, DC\",\"wos_standard\":\"WEBER, DC\",\"first_name\":\"DC\",\"last_name\":\"WEBER\"}],\"pub_info\":{\"issue\":\"5\",\"pubtype\":\"Journal\",\"sortdate\":\"1972-01-01\",\"has_abstract\":\"N\",\"coverdate\":\"1972\",\"vol\":\"33\",\"pubyear\":\"1972\",\"page\":{\"end\":\"413\",\"page_count\":\"1\",\"begin\":\"413\"}},\"publishers\":[{\"addr_no\":\"1\",\"role\":\"publisher\",\"seq_no\":\"1\",\"display_name\":\"ASSOC COLL RESEARCH LIBRARIES\",\"full_name\":\"ASSOC COLL RESEARCH LIBRARIES\",\"address\":{\"addr_no\":\"1\",\"full_address\":\"50 E HURON ST, CHICAGO, IL 60611\",\"city\":\"CHICAGO\"}}],\"titles\":{\"source\":\"COLLEGE \\u0026 RESEARCH LIBRARIES\",\"source_abbrev\":\"COLL RES LIBR\",\"abbrev_iso\":\"Coll. Res. Libr.\",\"abbrev_11\":\"COLL RES LI\",\"abbrev_29\":\"COLL RES LIBR\",\"item\":\"LIBRARY MANAGEMENT - BEHAVIOR-BASED PERSONNEL SYSTEMS (BBPS) - FRAMEWORK FOR ANALYSIS - KEMPER,RE\"}}"
```

The `WosRecord#summary_struct` recursively applies OpenStruct, e.g.
```ruby
# work with an OpenStruct that is applied recursively to the summary Hash
> s = wos_record_encoded.summary_struct;
> s.titles.item
=> "LIBRARY MANAGEMENT - BEHAVIOR-BASED PERSONNEL SYSTEMS (BBPS) - FRAMEWORK FOR ANALYSIS - KEMPER,RE"
> s.names.first.last_name
=> "WEBER"
> s.names.map {|n| n.last_name }
=> ["WEBER"]
```

### CODE REVIEW

**Comment by atz** [comment]
@darrenleeweber Are you opening to squashing some or all of this?

**Comment by darrenleeweber** [comment]
Squashed all the `WosRecord` commits.

File: lib/wos_record.rb [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+# Utilities for working with a Web of Science record
+class WosRecord
+
+  # @return doc [Nokogiri::XML::Document] WOS record document
+  attr_reader :doc
+
+  # @param record [String] record in XML
+  # @param encoded_record [String] record in HTML encoding
+  def initialize(record: nil, encoded_record: nil)
```
Comment (atz) [inline lib/wos_record.rb]: How about a some protection, like:
```ruby
raise ArgumentError, 'Either record or encoded_record parameter is required' if record.nil? && encoded_Record.nil?
```

Comment (darrenleeweber) [reply lib/wos_record.rb]: Yes, for sure.  It's already in there, see the private method.  It's OK if `record` is nil, but when it is, the `encoded_record` cannot be nil.  The specs check that this raises an exception (IIRC).

Comment (darrenleeweber) [reply lib/wos_record.rb]: See https://github.com/sul-dlss/sul_pub/pull/275/files#diff-3062786315e5c2629414d0c5972f1caaR150

**Comment by atz** [comment]
OK, thanks for that clarification.

**Comment by Maatary** [comment]
A first glimpse at the actual first step of the mapping from science wire to a pub_hash (this is not the complete set of information but the bulk)

in the ScienceWireHarvester
```
pub_hash = SciencewireSourceRecord.convert_sw_publication_doc_to_hash(incoming_sw_xml_doc)
```

in SciencewireSourceRecord

```
 def self.convert_sw_publication_doc_to_hash(publication)
    doi = extract_doi(publication)
    issn = extract_issn(publication)
    pmid = extract_pmid(publication)
    swid = extract_swid(publication)
    wosid = extract_wosid(publication)

    doi_identifier = { type: 'doi', id: doi, url: "#{Settings.DOI.BASE_URI}#{doi}" } unless doi.blank?
    issn_identifier = { type: 'issn', id: issn, url: Settings.SULPUB_ID.SEARCHWORKS_URI + issn } unless issn.blank?

    identifiers = []
    identifiers << { type: 'PMID', id: pmid, url: "#{Settings.PUBMED.ARTICLE_BASE_URI}#{pmid}" } unless pmid.blank?
    identifiers << { type: 'WoSItemID', id: wosid, url: "#{Settings.SCIENCEWIRE.ARTICLE_BASE_URI}#{wosid}" } unless wosid.blank?
    identifiers << { type: 'PublicationItemID', id: swid } unless swid.blank?
    identifiers << doi_identifier unless doi.blank?

    record_as_hash = {}
    record_as_hash[:provenance] = Settings.sciencewire_source
    record_as_hash[:sw_id] = swid
    record_as_hash[:pmid] = pmid unless pmid.blank?
    record_as_hash[:issn] = issn unless issn.blank?
    record_as_hash[:identifier] = identifiers

    record_as_hash[:title] = publication.xpath('Title').text unless publication.xpath('Title').blank?
    record_as_hash[:abstract_restricted] = publication.xpath('Abstract').text unless publication.xpath('Abstract').blank?
    record_as_hash[:author] = publication.xpath('AuthorList').text.split('|').collect { |author| { name: author } }

    record_as_hash[:year] = publication.xpath('PublicationYear').text unless publication.xpath('PublicationYear').blank?
    record_as_hash[:date] = publication.xpath('PublicationDate').text unless publication.xpath('PublicationDate').blank?

    record_as_hash[:authorcount] = publication.xpath('AuthorCount').text unless publication.xpath('AuthorCount').blank?

    record_as_hash[:keywords_sw] = publication.xpath('KeywordList').text.split('|') unless publication.xpath('KeywordList').blank?
    record_as_hash[:documenttypes_sw] = publication.xpath('DocumentTypeList').text.split('|')

    record_as_hash[:documentcategory_sw] = publication.xpath('DocumentCategory').text unless publication.xpath('DocumentCategory').blank?
    sul_document_type = lookup_cap_doc_type_by_sw_doc_category(record_as_hash[:documentcategory_sw])
    record_as_hash[:type] = sul_document_type

    record_as_hash[:publicationimpactfactorlist_sw] = publication.xpath('PublicationImpactFactorList').text.split('|') unless publication.xpath('PublicationImpactFactorList').blank?
    record_as_hash[:publicationcategoryrankinglist_sw] = publication.xpath('PublicationCategoryRankingList').text.split('|') unless publication.xpath('PublicationCategoryRankingList').blank?
    record_as_hash[:numberofreferences_sw] = publication.xpath('NumberOfReferences').text unless publication.xpath('NumberOfReferences').blank?
    record_as_hash[:timescited_sw_retricted] = publication.xpath('TimesCited').text unless publication.xpath('TimesCited').blank?
    record_as_hash[:timenotselfcited_sw] = publication.xpath('TimesNotSelfCited').text unless publication.xpath('TimesNotSelfCited').blank?
    record_as_hash[:authorcitationcountlist_sw] = publication.xpath('AuthorCitationCountList').text unless publication.xpath('AuthorCitationCountList').blank?
    record_as_hash[:rank_sw] = publication.xpath('Rank').text unless publication.xpath('Rank').blank?
    record_as_hash[:ordinalrank_sw] = publication.xpath('OrdinalRank').text unless publication.xpath('OrdinalRank').blank?
    record_as_hash[:normalizedrank_sw] = publication.xpath('NormalizedRank').text unless publication.xpath('NormalizedRank').blank?
    record_as_hash[:newpublicationid_sw] = publication.xpath('NewPublicationItemID').text unless publication.xpath('NewPublicationItemID').blank?
    record_as_hash[:isobsolete_sw] = publication.xpath('IsObsolete').text unless publication.xpath('IsObsolete').blank?

    record_as_hash[:publisher] = publication.xpath('CopyrightPublisher').text unless publication.xpath('CopyrightPublisher').blank?
    record_as_hash[:city] = publication.xpath('CopyrightCity').text unless publication.xpath('CopyrightCity').blank?
    record_as_hash[:stateprovince] = publication.xpath('CopyrightStateProvince').text unless publication.xpath('CopyrightStateProvince').blank?
    record_as_hash[:country] = publication.xpath('CopyrightCountry').text unless publication.xpath('CopyrightCountry').blank?
    record_as_hash[:pages] = publication.xpath('Pagination').text unless publication.xpath('Pagination').blank?

    if sul_document_type == Settings.sul_doc_types.inproceedings
      conference_hash = {}
      conference_hash[:name] = publication.xpath('ConferenceTitle').text unless publication.xpath('ConferenceTitle').blank?
      conference_hash[:startdate] = publication.xpath('ConferenceStartDate').text unless publication.xpath('ConferenceStartDate').blank?
      conference_hash[:enddate] = publication.xpath('ConferenceEndDate').text unless publication.xpath('ConferenceEndDate').blank?
      conference_hash[:city] = publication.xpath('ConferenceCity').text unless publication.xpath('ConferenceCity').blank?
      conference_hash[:statecountry] = publication.xpath('ConferenceStateCountry').text unless publication.xpath('ConferenceStateCountry').blank?
      record_as_hash[:conference] = conference_hash unless conference_hash.empty?
    elsif sul_document_type == Settings.sul_doc_types.book
      record_as_hash[:booktitle] = publication.xpath('PublicationSourceTitle').text unless publication.xpath('PublicationSourceTitle').blank?
      record_as_hash[:pages] = publication.xpath('Pagination').text unless publication.xpath('Pagination').blank?
    end

    if sul_document_type == Settings.sul_doc_types.article || (sul_document_type == Settings.sul_doc_types.inproceedings && !publication.xpath('Issue').blank?)
      journal_hash = {}
      journal_hash[:name] = publication.xpath('PublicationSourceTitle').text unless publication.xpath('PublicationSourceTitle').blank?
      journal_hash[:volume] = publication.xpath('Volume').text unless publication.xpath('Volume').blank?
      journal_hash[:issue] = publication.xpath('Issue').text unless publication.xpath('Issue').blank?
      journal_hash[:articlenumber] = publication.xpath('ArticleNumber').text unless publication.xpath('ArticleNumber').blank?
      journal_hash[:pages] = publication.xpath('Pagination').text unless publication.xpath('Pagination').blank?
      journal_identifiers = []
      journal_identifiers << issn_identifier unless issn.blank?
      journal_identifiers << doi_identifier unless doi.blank?
      journal_hash[:identifier] = journal_identifiers
      record_as_hash[:journal] = journal_hash
    end

    # unless issn.blank? || publication.xpath('Issue').blank? || sul_document_type == Settings.sul_doc_types.article
    #   book_series_hash = {}
    #   book_series_hash[:identifier] = [issn_identifier]
    #   book_series_hash[:title] = publication.xpath('PublicationSourceTitle').text unless publication.xpath('PublicationSourceTitle').blank?
    #   book_series_hash[:volume] = publication.xpath('Volume').text unless publication.xpath('Volume').blank?
    #   record_as_hash[:series] = book_series_hash
    # end
    record_as_hash
  end
```

**Pull request merged**